protocol/x/subaccounts/keeper/isolated_subaccount.go

```diff
@@ -135,8 +186,148 @@ func isValidIsolatedPerpetualUpdates(
+			isolatedUpdate,
+		),
+		nil
+}
+
+// getIsolatedPerpetualStateTransition computes whether an isolated perpetual position will be
+// opened or closed for a subaccount given an isolated perpetual update for the subaccount.
+// Input subaccount account must be settled.
+func getIsolatedPerpetualStateTransition(
+	settledSubaccount types.Subaccount,
+	isolatedPerpetualPosition *types.PerpetualPosition,
+	isolatedPerpetualUpdate *types.PerpetualUpdate,
+) *isolatedPerpetualStateTransition {
+	// If there is no update to an isolated perpetual position, then no state transitions have
+	// happened for the isolated perpetual.
+	if isolatedPerpetualUpdate == nil {
+		return nil
+	}
+
+	perpetualId := isolatedPerpetualUpdate.PerpetualId
+	// TODO(DEC-715): Support non-USDC assets.
+	usdcQuantumsBeforeUpdate := new(big.Int).Set(settledSubaccount.GetUsdcPosition())
+
+	// If the subaccount has no isolated perpetual position, then this update is opening an isolated
+	// perpetual position.
+	if isolatedPerpetualPosition == nil {
+		return &isolatedPerpetualStateTransition{
+			perpetualId:              perpetualId,
+			usdcQuantumsBeforeUpdate: usdcQuantumsBeforeUpdate,
+			transition:               opened,
+		}
+	}
+
+	// If the position size after the update is zero, then this update is closing an isolated
+	// perpetual position.
+	if finalPositionSize := new(big.Int).Add(
+		isolatedPerpetualPosition.GetBigQuantums(),
+		isolatedPerpetualUpdate.GetBigQuantums(),
+	); finalPositionSize.Cmp(lib.BigInt0()) == 0 {
+		return &isolatedPerpetualStateTransition{
+			perpetualId:              perpetualId,
+			usdcQuantumsBeforeUpdate: usdcQuantumsBeforeUpdate,
+			transition:               closed,
+		}
+	}
+
+	// If a position was not opened or closed, no state transition happened from the perpetual
+	// update.
+	return nil
+}
+
+// transferCollateralForIsolatedPerpetual transfers collateral between an isolated collateral pool
+// and the cross-perpetual collateral pool based on whether an isolated perpetual position was
+// opened or closed in a subaccount.
+// Note: This uses the `x/bank` keeper and modifies `x/bank` state.
+func (k *Keeper) transferCollateralForIsolatedPerpetual(
+	ctx sdk.Context,
+	updatedSubaccount types.Subaccount,
+	stateTransition *isolatedPerpetualStateTransition,
+) error {
+	// No collateral to transfer if no state transition.
+	if stateTransition == nil {
+		return nil
+	}
+
+	isolatedCollateralPoolAddr, err := k.GetCollateralPoolFromPerpetualId(ctx, stateTransition.perpetualId)
+	if err != nil {
+		return err
+	}
+	var toModuleAddr sdk.AccAddress
+	var fromModuleAddr sdk.AccAddress
+	var usdcQuantums *big.Int
+
+	// If an isolated perpetual position was opened in the subaccount, then move collateral equivalent
+	// to the USDC asset position size of the subaccount before the update from the
+	// cross-perpetual collateral pool to the isolated perpetual collateral pool.
+	if stateTransition.transition == opened {
```

Comment:
Related to this [question](https://github.com/dydxprotocol/v4-chain/pull/1200/files#r1531458231), is there no automatic collateral transfer involved in increasing a position? Or is there a separate transfer flow to increase collateral?

Reply:
That is correct. Increasing / decreasing a position does not lead to automatic collateral transfer. To increase collateral for an isolated perpetual position, a user would have to transfer in additional collateral, the same way they would increase collateral for some set of cross-margined perpetual positions.